lib/plugins/wallet/coinbasepro/coinbasepro.js

```diff
@@ -0,0 +1,163 @@
+
+const API_URL = 'https://api.exchange.coinbase.com'
+
+function request (options, account) {
+  // create the json request object
+  const cbAccessTimestamp = Date.now() / 1000 // in ms
+  const method = options.method
+  const requestPath = options.path
+  const body = _.defaultTo('', JSON.stringify(options.body))
+
+  // create the prehash string by concatenating required parts
+  const message = cbAccessTimestamp + method + requestPath + body
+
+  // decode the base64 secret
+  const secret = Buffer.from(account.apiSecret, 'base64')
+
+  // create a sha256 hmac with the secret
+  const hmac = crypto.createHmac('sha256', secret)
+
+  // sign the require message with the hmac and base64 encode the result
+  const cbAccessSign = hmac.update(message).digest('base64')
+  const headers = {
+    'content-type': 'application/json',
+    'cb-access-key': account.apiKey,
+    'cb-access-sign': cbAccessSign,
+    'cb-access-timestamp': cbAccessTimestamp,
+    'cb-access-passphrase': account.passphrase
+  }
+  const config = {
+    baseURL: API_URL,
+    url: requestPath,
+    method,
+    headers,
+    body: options.body
+  }
+  return axios(config)
+    .then(res => res.data)
+    .catch(err => console.error('Error: ' + err.message))
+}
+
+function balance (account, cryptoCode, settings, operatorId) {
+  return checkCryptoCode(cryptoCode)
+    .then(() => getWallet(account, cryptoCode))
+    .then(_.get('balance'))
+}
+
+function newAddress (account, info, tx, settings, operatorId) {
+  return checkCryptoCode(info.cryptoCode)
+    .then(() => getWallet(account, info.cryptoCode))
+    .then(wallet => {
+      return request({
+        path: `/coinbase-accounts/${wallet.id}/addresses`,
+        method: 'POST'
+      }, account)
+        .then(_.get('address'))
+    })
+}
+
+function convertPendingBalance (ratesRec, balance, cryptoCode) { 
+      const rates = ratesRec.rates
+      const rate = (rates.ask.plus(rates.bid)).div(2)
+      const cryptoRec = coinUtils.getCryptoCurrency(cryptoCode)
+      const unitScale = cryptoRec.unitScale
+      const baseUnits = new BN(1).shiftedBy(unitScale)
+      const result = baseUnits.times(balance).div(rate)
+      return result
+}
+
+function getRates(fiatCode, cryptoCode) {
+  return settingsLoader.loadLatest().then(settings => ticker.getRates(settings, fiatCode, cryptoCode))
+}
+
+function newFunding (account, cryptoCode, fiatCode, settings, operatorId) {
+  return checkCryptoCode(cryptoCode)
+    .then(() => {
+      const promises = [
+        getWallet(account, cryptoCode),
+        newAddress(account, { cryptoCode }),
+        getRates(fiatCode, cryptoCode)
+      ]
+      return Promise.all(promises)
+    })
+    .then(([wallet, fundingAddress, ratesRec]) => ({
+      fundingConfirmedBalance: new BN(wallet.balance),
+      fundingPendingBalance: convertPendingBalance(ratesRec, new BN(wallet['hold_balance']), cryptoCode),
+      fundingAddress
+    }))
+}
```

Comment:
Not sure if I like having ticker related logic in a wallet file. How hard would it be to pass the rate as a parameter for `newFunding()`? If you choose this, make sure to rename the other parameters accordingly

```suggestion
function newFunding (account, cryptoCode, fiatRate, settings, operatorId) {
  return checkCryptoCode(cryptoCode)
    .then(() => {
      const promises = [
        getWallet(account, cryptoCode),
        newAddress(account, { cryptoCode }),
        fiatRate
      ]
      return Promise.all(promises)
    })
    .then(([wallet, fundingAddress, ratesRec]) => ({
      fundingConfirmedBalance: new BN(wallet.balance),
      fundingPendingBalance: convertPendingBalance(ratesRec, new BN(wallet['hold_balance']), cryptoCode),
      fundingAddress
    }))
}
```

Reply:
@RafaelTaranto any thoughts about this?